trl/trainer/callbacks.py

```diff
@@ -273,6 +278,30 @@ def on_train_begin(self, args: TrainingArguments, state: TrainerState, control:
                 generation_config=self.generation_config,
                 batch_size=args.per_device_eval_batch_size,
             )
+            # Compute initial win rate as a reference point
+            completions = list(zip(self.ref_completions, self.ref_completions))
+            winner_indices = self.judge.judge(prompts, completions, self.shuffle_order)
+            prompts = gather_object(prompts)
+            completions = gather_object(completions)
+            winner_indices = gather_object(winner_indices)
+
+        # Logging
+        if self.trainer.accelerator.is_main_process:
+            win_rate = sum(winner_idx == 1 for winner_idx in winner_indices) / len(winner_indices)
+            self.trainer.log({"eval_win_rate": win_rate})
+
+            if "wandb" in args.report_to:
+                import wandb
+
+                if wandb.run is not None:
+                    global_step = [str(state.global_step)] * len(prompts)
+                    data = list(zip(global_step, prompts, completions, winner_indices))
+                    # Split completions from referenece model and policy
+                    split_data = [(item[0], item[1], item[2][0], item[2][1], item[3]) for item in data]
+                    df = pd.DataFrame(
+                        split_data, columns=["step", "prompt", "reference_model", "policy", "winner_index"]
+                    )
+                    wandb.log({"completions": wandb.Table(dataframe=df)})
```

Comment:
Note that logging tables this way gives a nice step counter (unlike `LogCompletionsCallback`)

![Screenshot 2024-09-26 at 14 15 34](https://github.com/user-attachments/assets/77de2664-1350-45a6-ba8c-ac617afeb01f)